### PR TITLE
Perf optimization for site extension installation

### PR DESCRIPTION
**Pull request by shrimpy**
improve perf for the case that installation finished quick and doesn`t need to trigger a restart

this will also benefit for scenario that install an site extension where it is already installed, in this case installation will run synchronously

### CODE REVIEW

File: Kudu.Core/SiteExtensions/SiteExtensionManager.cs [hunk 1]
```diff
@@ -308,7 +314,7 @@ private SiteExtensionInfo GetPreInstalledExtension(string id)
             }
         }
 
-        private async Task<SiteExtensionInfo> TryInstallExtension(string id, string version, string feedUrl, SiteExtensionInfo.SiteExtensionType type, ITracer tracer)
+        private async Task<SiteExtensionInfo> TryInstallExtension(string id, string version, string feedUrl, SiteExtensionInfo.SiteExtensionType type, ITracer tracer, AutoResetEvent installationSignal = null)
```
Comment (davidebbo) [inline Kudu.Core/SiteExtensions/SiteExtensionManager.cs]: nitpick: `=null` default value not needed here

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs [hunk 2]
```diff
@@ -117,6 +119,11 @@ public bool IsTerminalStatus()
                 || string.Equals(Constants.SiteExtensionProvisioningStateCanceled, ProvisioningState, StringComparison.OrdinalIgnoreCase);
         }
 
+        public bool IsRequireRestart()
```
Comment (davidebbo) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: `IsRestartRequired` is a bit clearer :)

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs [hunk 3]
```diff
@@ -117,6 +119,11 @@ public bool IsTerminalStatus()
...
+        {
+            return IsSiteExtensionRequireRestart(this, _rootPath);
```
Comment (davidebbo) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: Why is `IsSiteExtensionRequireRestart` static? If it takes the instance in, why not make it instance?

File: Kudu.Core/SiteExtensions/SiteExtensionManager.cs (hunk 1 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionManager.cs]: this will be removed.

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs (hunk 2 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: sure. will update it.

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs (hunk 3 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: "IsSiteExtensionRequireRestart" is for shared the logic with "IsAnyInstallationRequireRestart"

when perform GET, we use "IsAnyInstallationRequireRestart" to check if there is any newly installed extension required to restart, if we find out, we will return header to GEO to trigger restart. and "IsAnyInstallationRequireRestart" is static, that is why "IsSiteExtensionRequireRestart" is static

File: Kudu.Contracts/SiteExtensions/ISiteExtensionManager.cs [hunk 4]
```diff
@@ -1,4 +1,6 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
```
Comment (shrimpy) [inline Kudu.Contracts/SiteExtensions/ISiteExtensionManager.cs]: reverted

**Pull request merged**